### PR TITLE
I am Spartacus

### PR DESCRIPTION
**Pull request by favna**
The content here is pretty outdated but i cba to update it.

-----

I spy with my little eye and I see TypeScript and meme commits so you got my interest 👀

Cleaned up the repo quite a bit, removing various non-needed stuff. Lets go over each file. If you have any questions please ask.

## Concerning `.eslintrc.js`
Eslint recently added the [`require-atomic-updates`](https://eslint.org/docs/rules/require-atomic-updates) rule to their recommended set and it has been plagueing my projects as well. It was throwing an errror in PlayerManager.ts for you so I disabled the rule entirely. You can read up on the rule there and decide if you want to re-enable it or put it to warn mode or something.

## Concerning the removal of `.npmignore`
I added the `files` field to package.json. It's far better to specify which files you *want* to include than which files you *do not want* to include as the latter changes while the former is constant. See NPM publishing sample for the difference.
  - the azure pipelines yaml is published to npm right now btw, just saying. Not that it really matters, you won't have to make this change on master and make a v2 release for that but just saying.

## Concerning `package.json`
Mostly see the section about rollup. Other than that I bumped ws which has a patch release and changed the peerDependency rule to show what it actually does (`^11.5.1` allows the exact same as `^11`). I also set the engines field, which gets verified by yarn (sadly not NPM, bad NPM).

## Concerning `docs.js`
I changed it to `docs.ts` and moved it to scripts. In `package.json` it is now called with `ts-node` which is a TypeScript execution and REPL for node.js package (that is to say, `node` but then for TS files)

## Concerning `tsconfig.json`
  - I reordered the keys a bit so they are grouped a bit easier
  - lib `esnext` implied everything esnext so there is no need to include the sub-sections
  - targeting es6 is better because far from everyone (in fact probably a minority with the djs "noobs" out there) use Node v10/11/12. While ESNext is not necessarily targetting those versions, run anything lower and you can run into problems (i.e. Array.flat and Array.flatMap only being on Node v11+ while being esnext)
  - You weren't using any decorators thus far and to be frank, I don't see this project needing any. I removed the option for now.
  - alwaysStrict, strictFunctionTypes and strictBindCallApply are implied by strict mode. They are not individually required when strict is true.
  - I changed "allowSyntheticDefaultImports" to "esModuleInterop", which implies "allowSyntheticDefaultImports". It allows for a bit more ease  when importing stuff than "allowSyntheticDefaultImports" does.
    - LavalinkNode.ts had its import of `import * as Websocket from 'ws'` edited to `import Websocket from 'ws'` because of this.
  - I added "forceConsistentCasingInFileNames", "noUnusedParameters", "noImplicitReturns", "noImplicitThis" and "noFallthroughCaseInSwitch", three more anti-bug features that are nice to have.
  - I removed resolveJsonModule because you're not using any JSON modules (it's for loading JSON with import statements)
  - pretty is always true, unless you're piping output to another program or redirecting output to a file. 
  - types node is redudant, just install them under your devDependencies like you did
  - exclusions are only relevant when you want to exclude files that are previously in your include. For example if you have a tests folder inside your src folder. I removed the option.

### Before remove npmignore

```
npm publish --dry-run

> discord.js-lavalink@3.0.0 prepublishOnly .
> yarn compile

yarn run v1.17.3
$ tsc -p .
✨  Done in 1.15s.
npm notice
npm notice package: discord.js-lavalink@3.0.0
npm notice === Tarball Contents === 
npm notice 1.3kB  package.json
npm notice 1.1kB  LICENSE
npm notice 4.3kB  README.md
npm notice 578B   tsconfig.json
npm notice 400B   .github/FUNDING.yml
npm notice 106B   dist/index.d.ts
npm notice 333B   dist/index.js
npm notice 144B   dist/index.js.map
npm notice 1.4kB  dist/lib/LavalinkNode.d.ts   
npm notice 3.9kB  dist/lib/LavalinkNode.js     
npm notice 4.3kB  dist/lib/LavalinkNode.js.map 
npm notice 1.7kB  dist/lib/Player.d.ts
npm notice 3.5kB  dist/lib/Player.js
npm notice 3.5kB  dist/lib/Player.js.map       
npm notice 2.0kB  dist/lib/PlayerManager.d.ts  
npm notice 5.2kB  dist/lib/PlayerManager.js    
npm notice 5.6kB  dist/lib/PlayerManager.js.map
npm notice 84.4kB dist/tsconfig.tsbuildinfo    
npm notice 103B   src/index.ts
npm notice 4.8kB  src/lib/LavalinkNode.ts      
npm notice 4.6kB  src/lib/Player.ts
npm notice 6.3kB  src/lib/PlayerManager.ts     
npm notice === Tarball Details === 
npm notice name:          discord.js-lavalink
npm notice version:       3.0.0
npm notice package size:  21.8 kB
npm notice unpacked size: 139.5 kB
npm notice shasum:        f26e2bcc0f76cf36b541d9288dcf2c7f1482af69
npm notice integrity:     sha512-OC4+WQraSAx+b[...]Sh0rgP/RmO5rg==
npm notice total files:   22
npm notice 
+ discord.js-lavalink@3.0.0
```

### After removing npmignore

```
npm publish --dry-run

> discord.js-lavalink@3.0.0 prepublishOnly .
> yarn compile

yarn run v1.17.3
$ tsc -p .
✨  Done in 1.16s.
npm notice
npm notice package: discord.js-lavalink@3.0.0
npm notice === Tarball Contents === 
npm notice 1.3kB  package.json
npm notice 1.1kB  LICENSE
npm notice 4.3kB  README.md
npm notice 106B   dist/index.d.ts
npm notice 333B   dist/index.js
npm notice 144B   dist/index.js.map
npm notice 1.4kB  dist/lib/LavalinkNode.d.ts   
npm notice 3.9kB  dist/lib/LavalinkNode.js     
npm notice 4.3kB  dist/lib/LavalinkNode.js.map 
npm notice 1.7kB  dist/lib/Player.d.ts
npm notice 3.5kB  dist/lib/Player.js
npm notice 3.5kB  dist/lib/Player.js.map       
npm notice 2.0kB  dist/lib/PlayerManager.d.ts  
npm notice 5.2kB  dist/lib/PlayerManager.js    
npm notice 5.6kB  dist/lib/PlayerManager.js.map
npm notice 84.4kB dist/tsconfig.tsbuildinfo    
npm notice === Tarball Details === 
npm notice name:          discord.js-lavalink
npm notice version:       3.0.0
npm notice package size:  17.4 kB
npm notice unpacked size: 122.7 kB
npm notice shasum:        90bd73ec42a789f9ea1a14daa4ecbc4e4da43e75
npm notice integrity:     sha512-NQLUVkLLlHDwe[...]XPAsInXbtnh6A==
npm notice total files:   16
npm notice 
+ discord.js-lavalink@3.0.0
```

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -1,37 +1,24 @@
...
 		"removeComments": true,
-		"resolveJsonModule": true,
+		"declaration": true,
 		"sourceMap": true,
-		"strictFunctionTypes": true,
-		"strictBindCallApply": true,
-		"target": "esnext",
-		"types": ["node"]
 	},
-	"exclude": [
```
Comment (MrJacz) [inline tsconfig.json:24]: why was this removed

File: package.json [hunk 2]
```diff
@@ -35,19 +38,39 @@
...
   },
   "peerDependencies": {
-    "discord.js": "^11.5.1"
+    "discord.js": "^11"
+  },
+  "files": [
+    "dist"
+  ],
+  "engines": {
+    "node": ">=8",
```
Comment (MrJacz) [inline package.json]: node is actually >=10

File: package.json [hunk 3]
```diff
@@ -35,19 +38,39 @@
...
   },
   "dependencies": {
-    "ws": "^7.1.1"
+    "ws": "^7.1.2"
   },
   "peerDependencies": {
-    "discord.js": "^11.5.1"
+    "discord.js": "^11"
+  },
+  "files": [
```
Comment (MrJacz) [inline package.json]: I would like to include the LICENCE and readme

**Review by MrJacz — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 1 above)
Comment (favna) [reply tsconfig.json:24]: Exclusions are only relevant when you want to exclude something that was previously included. i.e. if you have the "tests" folder under "src" or keep some legacy code in there. 

TL;DR: Unless you need to filter what is included, exclusions aren't necessary.

File: package.json (hunk 3 above)
Comment (favna) [reply package.json]: License, package, readme and changelog are always included. Even if you add them to npmignore. And they're still included now (just see the second dry run)

Comment (favna) [reply package.json]: Sidenote about that dry run, I see it holds a TS buildinfo file. It didn't do that in later dry runs. Not sure why it landed in dist before.

File: package.json (hunk 2 above)
Comment (favna) [reply package.json]: K, I'll change this

File: package.json [hunk 4]
```diff
@@ -2,15 +2,17 @@
   "name": "discord.js-lavalink",
   "version": "3.0.0",
   "description": "A discord.js lavalink client",
-  "main": "dist/index",
+  "main": "dist/index.js",
+  "module": "dist/index.es.js",
```
Comment (MrJacz) [inline package.json]: remove this

File: package.json [hunk 5]
```diff
@@ -2,15 +2,17 @@
...
+  "jsnext:main": "dist/index.es.js",
```
Comment (MrJacz) [inline package.json]: this 2

File: rollup.config.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import commonjs from "rollup-plugin-commonjs";
```
Comment (MrJacz) [inline rollup.config.ts]: remove this entire file

**Review by MrJacz — CHANGES_REQUESTED** [review]

File: tsconfig.json [hunk 7]
```diff
@@ -1,37 +1,24 @@
...
+		"lib": ["esnext",],
+		"target": "es6",
 		"moduleResolution": "node",
-		"noUnusedLocals": true,
-		"outDir": "./dist",
-		"pretty": true,
+		"esModuleInterop": true,
+		"strict": true,
+		"noImplicitReturns": true,
+        "noImplicitThis": true,
```
Comment (MrJacz) [inline tsconfig.json]: fix indent please

File: tsconfig.json [hunk 8]
```diff
@@ -1,37 +1,24 @@
...
+        "noUnusedLocals": true,
```
Comment (MrJacz) [inline tsconfig.json:14]: same here

File: tsconfig.json [hunk 9]
```diff
@@ -1,37 +1,24 @@
...
-			"esnext.array",
-			"esnext.asynciterable",
-			"esnext.intl",
-			"esnext.symbol"
-		],
+		"outDir": "dist",
+		"rootDir": "src",
 		"module": "commonjs",
+		"lib": ["esnext",],
+		"target": "es6",
```
Comment (MrJacz) [inline tsconfig.json]: esnext please

**Comment by favna** [comment]
oi @MrJacz stop slacking

File: src/lib/LavalinkNode.ts [hunk 10]
```diff
@@ -113,7 +113,12 @@ export class LavalinkNode {
         });
     }
 
-    public configureResuming(key: string = Math.random().toString(36).substring(7), timeout = 120): Promise<boolean> {
+    public configureResuming(
+        key = Math.random()
```
Comment (MrJacz) [inline src/lib/LavalinkNode.ts:117]: no

**Review by MrJacz — CHANGES_REQUESTED** [review]

**Comment by MrJacz** [comment]
@Favna Can you fix these issues then i'll merge?
everything is setup for workflows
